[RHOAIENG-7942]: Update Side Panel Behavior to Pan Graph on Node Selection

jenny-s51 opened the pull request:
<!--- If this is a non-code change, this template is not required; reference any issues or top-level descriptions as needed -->
<!--- All code change PRs should relate to an issue, reference it here; see example below -->
<!--- Closes: #123 -->
https://issues.redhat.com/browse/RHOAIENG-7942

## Description
<!--- Describe your changes in detail; the what, the why, any findings, etc -->
<!--- Include any screenshots of changed UI; Include any gifs if it was a flow / UX change -->

Previously if we selected a pipelines node and the side panel opened, the node that toggled the panel was obscured by the panel.

Now when a node is selected, the node that toggled the drawer is panned into view. 

https://github.com/opendatahub-io/odh-dashboard/assets/32821331/e8bd847a-58e9-4958-b5f0-b8b9ce1d6019


## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

1. Navigate to a pipeline, pipeline run, or scheduled run
2. Click + drag the graph to the right of the topology view (such as to be obscured by the side panel opening on the right)
3. Select a node in the pipeline
4. Verify the node/graph pans to the left of the drawer panel and is not obscured by the panel

## Test Impact
<!--- What tests have you done to covert implemented functionality -->
<!--- If tests are not applicable, explain why here -->
N/A

## Request review criteria:
<!--- This PR will be merged by any repository approver when it meets all the points in the checklist -->
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
Self checklist (all need to be checked):
- [x] The developer has manually tested the changes and verified that the changes work
- [x] Commits have been squashed into descriptive, self-contained units of work (e.g. 'WIP' and 'Implements feedback' style messages have been removed)
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has added tests or explained why testing cannot be added (unit or cypress tests for related changes)

If you have UI changes: 
<!--- You can ignore these if you are doing manifest, backend, internal logic, etc changes; aka non-UI / visual changes -->
- [x] Included any necessary screenshots or gifs if it was a UI change.
- [x] Included tags to the UX team if it was a UI/UX change (find relevant UX in the [SMEs](https://github.com/opendatahub-io/odh-dashboard/tree/main/docs/smes.md) section).

After the PR is posted & before it merges:
- [x] The developer has tested their solution on a cluster by using the image produced by the PR to `main`

Gkrumbach07 commented on frontend/src/app/App.scss:
```diff
@@ -41,6 +41,12 @@ body,
       inset-inline-start: 12px;
     }
   }
+  .pf-topology-container {
+    overflow-y: hidden;
+    .pf-v5-c-drawer__panel.pf-m-resizable {
+      min-width: 350px;
+    }
+  }
```
Is this needed? doesnt the drawer already have a min width

jenny-s51 replied:
Removed this styling

Gkrumbach07 commented on frontend/src/concepts/pipelines/content/pipelinesDetails/pipeline/PipelineDetails.tsx:
```diff
@@ -201,54 +201,50 @@ const PipelineDetails: PipelineCoreDetailsPageComponent = ({ breadcrumbPath }) =
+                className="pf-v5-u-h-100"
+              >
+                <TabContentBody hasPadding className="pf-v5-u-h-100">
+                  <PipelineDetailsYAML
```
this yaml tab has a very small height now

jenny-s51 replied:
Thanks @Gkrumbach07 nice catch, fixed

Gkrumbach07 commented on frontend/src/concepts/pipelines/content/pipelinesDetails/pipeline/PipelineDetails.tsx:
```diff
@@ -82,92 +80,96 @@ const PipelineDetails: PipelineCoreDetailsPageComponent = ({ breadcrumbPath }) =
+          )
+        }
+      >
+        <PageSection
+          isFilled
+          padding={{ default: 'noPadding' }}
+          style={{ flexBasis: 0, overflowY: 'hidden' }}
+          variant="light"
+        >
+          <Flex
+            direction={{ default: 'column' }}
+            style={{ height: '100%' }}
+            spaceItems={{ default: 'spaceItemsNone' }}
+          >
+            <FlexItem>
```
this flex and page section dont do anything

Gkrumbach07 replied:
same thing on the run details too

jenny-s51 replied:
Thank you @Gkrumbach07 removed the `Flex` components; `PageSection` is needed for the tabs to persist when the drawer is opened

jeff-phillips-18 replied:
These Flex components are needed in order to size the graph area correctly such that the side panel scrolls and the graph doesn't expand below the height of the window which then puts the graph controls off screen.

Gkrumbach07 commented on frontend/src/concepts/pipelines/content/pipelinesDetails/pipeline/PipelineDetails.tsx:
```diff
@@ -82,92 +80,96 @@ const PipelineDetails: PipelineCoreDetailsPageComponent = ({ breadcrumbPath }) =
     );
   }
 
+  const panelContent = selectedNode ? (
+    <SelectedTaskDrawerContent
+      task={selectedNode.data.pipelineTask}
+      onClose={() => setSelectedId(null)}
+    />
+  ) : null;
+
```
why make this conditional. it looks like you can set if the panel is open in the viz surface

jenny-s51 replied:
Updated - thank you @Gkrumbach07 for your help this afternoon in resolving this

Gkrumbach07 commented on frontend/src/concepts/topology/PipelineVisualizationSurface.tsx:
```diff
@@ -25,14 +26,37 @@ import { NODE_HEIGHT, NODE_WIDTH } from './const';
+      const selectedNode = controller.getNodeById(selectedIds[0]);
+      if (selectedNode) {
+        // Use a timeout in order to allow the side panel to be shown and window size recomputed
+        resizeTimeout = setTimeout(() => {
+          controller.getGraph().panIntoView(selectedNode, { offset: 20, minimumVisible: 100 });
+          resizeTimeout = null;
+        }, 500);
```
it looks like there is a callback we can use in the `TopologyView`

```tsx
  /** Callback for side bar resize end. */
  onSideBarResize?: (width: number, id: string) => void;
```

jenny-s51 replied:
We found that this callback only fires when the sidebar is resized manually and not on the initial render

Gkrumbach07 commented:
![Screenshot 2024-07-01 at 7 59 20 AM](https://github.com/opendatahub-io/odh-dashboard/assets/12587674/9096e8de-6566-471f-9ee9-acbcb77e6b62)

@jenny-s51 It looks like the topology action buttons are pushed under the screen

jeff-phillips-18 commented:
With out the setting of the min-width for the side panel, we can get:
![image](https://github.com/opendatahub-io/odh-dashboard/assets/11633780/2af22a1a-435f-40da-b0b5-63cf632e7a04)

jeff-phillips-18 commented:
> @jenny-s51 It looks like the topology action buttons are pushed under the screen

This was the reason for:
```
  .pf-topology-container {
    overflow-y: hidden;
```

jenny-s51 reviewed (commented):
Thank you for your PR feedback @jeff-phillips-18 . Reverted to apply the styles and Flex components

Gkrumbach07 commented:
@jenny-s51 The drawer can be too small now and some tabs get hidden
![Screenshot 2024-07-02 at 5 24 09 PM](https://github.com/opendatahub-io/odh-dashboard/assets/12587674/853abeda-9d9c-44da-8bae-0be08f953984)

jenny-s51 commented:
@Gkrumbach07 while looking into the issue you found, I noticed a bug with the drawer component that was causing the drawer panel to remain in the view when the drawer was closed.  
![Screenshot 2024-07-03 at 11 30 23 AM](https://github.com/opendatahub-io/odh-dashboard/assets/32821331/f5a68bca-3bf3-4b4a-a9a6-f97b3a3dc0da)

@jeff-phillips-18 helped come up with a good solution to this, updated the PR with the fix. 

@Gkrumbach07 still seeing the issue with the hidden tabs though on initial resize. Investigating... this may be an issue on the PF side

jeff-phillips-18 commented:
> still seeing the issue with the hidden tabs though on initial resize. Investigating... this may be an issue on the PF side

Looks to be a PF issue. The Tabs component only determines if the scrolling components are needed when the window resizes. If the user resizes the side panel, the Tabs will not realize that the scroll components have become necessary.

jeff-phillips-18 commented on frontend/src/concepts/topology/PipelineVisualizationSurface.scss:
```diff
@@ -0,0 +1,8 @@
+  .pf-topology-container {
+    overflow-y: hidden;
+    .pf-v5-c-drawer__panel.pf-m-resizable {
+      min-width: 350px;
```
Previous version set the minimum with of the side panel to 500px, this fixes the hidden tab issue.
```suggestion
      min-width: 500px;
```

jenny-s51 replied:
Thank you for this awesome suggestion @jeff-phillips-18 🎉

jenny-s51 commented:
Investigating test errors in pipelinesTopology with the latest rebase/updates (after these files were recently refactored on `main`). It appears that some styling is missing after the rebase. I'm currently looking into this - will implement the necessary fixes and update accordingly.

jenny-s51 commented:
Updated/rebased with latest changes from `main` and fixed tests. 

@jeff-phillips-18 @Gkrumbach07 if you have a moment to take another look would be much appreciated

Gkrumbach07 reviewed (approved):
/lgtm
